Haskell-Generate/GenRustJets.hs

```diff
@@ -0,0 +1,346 @@
+    ["impl Jet for" <+> pretty modname <+> "{"
+    , rustJetCmr mod
+    , rustJetSourceTy mod
+    , rustJetTargetTy mod
+    , encodeStub
+    , decodeStub
+    , rustJetPtr mod
+    ])
+  , "}"
+  ]
+ where
+  modname = rustModuleName mod
+  encodeStub = vsep 
+    [ "fn encode<W: Write>(&self, w: &mut BitWriter<W>) -> std::io::Result<usize> {"
+    , "    todo!()"
+    , "}"
+    ]
+  decodeStub = vsep
+    [ "fn decode<I: Iterator<Item = u8>>(bits: &mut BitIter<I>) -> Result<Self, Error> {"
+    , "    todo!()"
+    , "}"
+    ]
+
+rustJetEnum :: Module -> Doc a
+rustJetEnum mod = vsep
+ [ pretty $ "/// " ++ rustModuleName mod ++ " jet family"
+ , "#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Hash)]"
+ , nest 4 (vsep (pretty ("pub enum " ++ rustModuleName mod ++ " {")
+   : [pretty (jetName jet) <> comma | (SomeArrow jet) <- moduleJets mod]))
+ , "}"
+ ]
+
+rustHeader :: Doc a
+rustHeader = "/* This file has been automatically generated. */"
+
+rustImports :: Module -> Doc a
+rustImports mod = vsep (map (<> semi)
+  [ "use crate::bititer::BitIter"
+  , "use crate::bitwriter::BitWriter"
+  -- , "use crate::jet::elements::ElementsEnv"
+  , "use crate::jet::type_name::TypeName"
+  , "use crate::jet::Jet"
+  , "use crate::merkle::cmr::Cmr"
+  , "use crate::{decode_bits, Error}"
+  , "use bitcoin_hashes::sha256::Midstate"
+  , "use simplicity_sys::c_jets::CTxEnv"
+  , "use simplicity_sys::CFrameItem"
+  , "use std::io::Write"
+  ])
+
+rustJetDoc :: Module -> SimpleDocStream a
+rustJetDoc mod = layoutPretty layoutOptions $ vsep (map (<> line)
+  [ rustHeader
+  , rustImports mod
+  , rustJetEnum mod
+  , rustJetImpl mod
+  ])
+
+rustFFIImports :: Doc a
+rustFFIImports = vsep (map (<> semi)
+  [ "use std::ffi::c_void"
+  , "use super::c_env::CElementsTxEnv"
+  , "use super::frame_ffi::CFrameItem"
+  ])
+
+rustFFISigs :: Module -> Doc a
+rustFFISigs mod = vsep
+  [ nest 4 (vsep ("extern \"C\" {": ((<> semi) . pretty . sig <$> moduleJets mod)))
+  , "}"
+  ]
+ where
+  sig (SomeArrow jet) = "pub fn "++cJetName jet++"(dst: *mut CFrameItem, src: CFrameItem, env: *const "++envType++") -> bool"
+   where
+    envType | CoreModule <- jetModule jet = "c_void"
+            | ElementsModule <- jetModule jet = "CElementsTxEnv"
+
+rustFFIDoc :: Module -> SimpleDocStream a
+rustFFIDoc mod = layoutPretty layoutOptions $ vsep (map (<> line)
+  [ rustHeader
+  , rustFFIImports
+  , rustFFISigs mod
+  ])
+
+rustWrapperImports :: Doc a
+rustWrapperImports = vsep (map (<> semi)
+  [ "use super::{frame_ffi::CFrameItem, jets_ffi, CTxEnv}"
+  ])
+
+rustWrappers :: Module -> Doc a
+rustWrappers mod = vsep ((<> line) . wrapper <$> moduleJets mod)
+ where
+  wrapper (SomeArrow jet) = vsep
+   [ nest 4 $ vsep
+     [ pretty $ "pub fn "++cJetName jet++"(dst: &mut CFrameItem, src: CFrameItem, "++envParam++") -> bool {"
+     , pretty $ "unsafe { "++lowerRustModuleName mod++"_ffi::"++cJetName jet++"(dst, src, "++envArg++") }"
```

Comment:
should be `jets_ffi::` instead of `elements_ffi::`

Reply:
Given the application specific nature of core jets, it makes sense to have `elements_ffi` instead of `jets_ffi`. Will update rust code to reflect that